File: years/2019/elixir/lib/aoc/day_01.ex
defmodule AoC.Day01 do
  @moduledoc false

  def part_1 do
    "../data/01.txt"
    |> File.stream!()
    |> Enum.map(&String.trim/1)
    |> Enum.map(&String.to_integer/1)
    |> Enum.map(&calculate_fuel/1)
    |> Enum.reduce(0, &(&1 + &2))
  end

  def part_2 do
    "../data/01.txt"
    |> File.stream!()
    |> Enum.map(&String.trim/1)
    |> Enum.map(&String.to_integer/1)
    |> Enum.map(&calculate_all_fuel/1)
    |> Enum.reduce(0, &(&1 + &2))
  end

  def calculate_fuel(mass), do: Integer.floor_div(mass, 3) - 2

  def calculate_all_fuel(mass) do
    case calculate_fuel(mass) do
      fuel_mass when fuel_mass > 0 -> fuel_mass + calculate_all_fuel(fuel_mass)
      _ -> 0
    end
  end
end

File: years/2019/swift/Sources/AoC/Day01.swift
/******************************************************************************
 **
 ** Copyright (c) 2023-2024 Kris Cottingham
 ** Licensed under the MIT License.
 **
 ** See https://github.com/krisalyssa/swift-aoc-common/blob/main/LICENSE
 ** for license information.
 **
 **/

import Common
import Foundation

public class Day01: Day {
  public init() {}

  public func part1(_ input: Input) {
    let data = input.asIntArray()
    let total = data.map { Day01.calculateFuel(mass: $0) }.reduce(0, +)

    print("day 01 part 1: \(total)")
  }

  public func part2(_ input: Input) {
    let data = input.asIntArray()
    let total = data.map { Day01.calculateAllFuel(mass: $0) }.reduce(0, +)

    print("day 01 part 2: \(total)")
  }

  static func calculateFuel(mass: Int) -> Int {
    return max((mass / 3) - 2, 0)
  }

  static func calculateAllFuel(mass: Int) -> Int {
    let fuelMass = calculateFuel(mass: mass)
    if fuelMass > 0 {
      return fuelMass + calculateAllFuel(mass: fuelMass)
    } else {
      return fuelMass
    }
  }
}
